File: gtci/07-stacks/cpp/2-reverse-a-string.cpp
#include <cassert>
#include <iostream>
#include <stack>
#include <string>

using namespace std;

class Solution {
   public:
    string reverseString(string s) {
        std::stack<char> stack;

        for (char ch : s) stack.push(ch);

        std::string reversedStr;
        while (stack.size() > 0) {
            reversedStr += stack.top();
            stack.pop();
        }

        return reversedStr;
    }
};

int main() {
    Solution sol;

    assert(sol.reverseString("hello") == "olleh");
    assert(sol.reverseString("world") == "dlrow");
    assert(sol.reverseString("Python") == "nohtyP");

    cout << "All test cases passed." << endl;

    return 0;
}

File: gtci/07-stacks/python3/2-reverse-a-string.py
class Solution:
    def reverseString(self, s):
        stack = []

        for char in s:
            stack.append(char)

        reversed_s = ""

        while len(stack) > 0:
            reversed_s += stack.pop()

        return reversed_s


if __name__ == "__main__":
    sol = Solution()

    assert sol.reverseString("hello") == "olleh"
    assert sol.reverseString("world") == "dlrow"
    assert sol.reverseString("Python") == "nohtyP"

    print("All test cases passed.")
